Extract all the various DB update calls and processes for Dag Parsing to one place

dimberman opened the pull request:
The updates to the various DB tables as a result of DAG file parsing and collection are scattered throughout the code base.

This change gathers them all in to one place to make it easier to a) find, and b) see what is changed.

The current situation is the result of organic code growth, but it is now almost impossible to find all the things we update as they are not organized, and were scatterd across at least DagBag, DagProcessor.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

ashb committed:
Extract all the various DB update calls and processes for Dag Parsing to one place

The updates to the various DB tables as a result of DAG file parsing and
collection are scattered throughout the code base.

This change gathers them all in to one place to make it easier to a) find, and
b) see what is changed.

The current situation is the result of organic code growth, but it is now
almost impossible to find all the things we update as they are not organized,
and were scatterd across at least DagBag, DagProcessor.

Co-authored-by: Daniel Imberman <daniel.imberman@gmail.com>

ashb commented on airflow/dag_processing/processor.py:
```diff
@@ -417,113 +419,8 @@ def __init__(self, dag_directory: str, log: logging.Logger):
 
-    @staticmethod
-    @provide_session
-    def update_import_errors(
```
Moved into collection

ashb commented on airflow/dag_processing/processor.py:
```diff
@@ -417,113 +419,8 @@ def __init__(self, dag_directory: str, log: logging.Logger):
-
-    @classmethod
-    @provide_session
-    def update_dag_warnings(cla, *, dagbag: DagBag, session: Session = NEW_SESSION) -> None:
```
Split, generating the warnings moved on to dagbag, writing moved to collection

ashb commented on tests/dag_processing/test_processor.py:
```diff
@@ -246,183 +243,6 @@ def test_process_file_should_failure_callback(self, monkeypatch, tmp_path, get_t
         msg = " ".join([str(k) for k in ti.key.primary]) + " fired callback"
         assert msg in callback_file.read_text()
 
-    @conf_vars({("core", "dagbag_import_error_tracebacks"): "False"})
-    def test_add_unparseable_file_before_sched_start_creates_import_error(self, tmp_path):
-        unparseable_filename = tmp_path.joinpath(TEMP_DAG_FILENAME).as_posix()
-        with open(unparseable_filename, "w") as unparseable_file:
-            unparseable_file.writelines(UNPARSEABLE_DAG_FILE_CONTENTS)
-
-        with create_session() as session:
-            self._process_file(unparseable_filename, dag_directory=tmp_path, session=session)
-            import_errors = session.query(ParseImportError).all()
-
-            assert len(import_errors) == 1
-            import_error = import_errors[0]
-            assert import_error.filename == unparseable_filename
-            assert import_error.stacktrace == f"invalid syntax ({TEMP_DAG_FILENAME}, line 1)"
-            session.rollback()
-
-    @conf_vars({("core", "dagbag_import_error_tracebacks"): "False"})
-    def test_add_unparseable_zip_file_creates_import_error(self, tmp_path):
-        zip_filename = (tmp_path / "test_zip.zip").as_posix()
-        invalid_dag_filename = os.path.join(zip_filename, TEMP_DAG_FILENAME)
-        with ZipFile(zip_filename, "w") as zip_file:
-            zip_file.writestr(TEMP_DAG_FILENAME, UNPARSEABLE_DAG_FILE_CONTENTS)
-
-        with create_session() as session:
-            self._process_file(zip_filename, dag_directory=tmp_path, session=session)
-            import_errors = session.query(ParseImportError).all()
-
-            assert len(import_errors) == 1
-            import_error = import_errors[0]
-            assert import_error.filename == invalid_dag_filename
-            assert import_error.stacktrace == f"invalid syntax ({TEMP_DAG_FILENAME}, line 1)"
-            session.rollback()
```
It could be worth keeping these to ensure that `dagbag.import_errors` is populated. Thoughts?

ashb replied:
Added back now.

kaxil replied:
Yea worth keeping it

ashb commented on tests/dag_processing/test_processor.py:
```diff
@@ -246,183 +243,6 @@ def test_process_file_should_failure_callback(self, monkeypatch, tmp_path, get_t
-            session.rollback()
-
-    @conf_vars({("core", "dagbag_import_error_tracebacks"): "False"})
-    def test_dag_model_has_import_error_is_true_when_import_error_exists(self, tmp_path, session):
```
Moved to test_collection.

ashb commented on tests/models/test_dagbag.py:
```diff
@@ -655,154 +607,6 @@ def f():
-    @patch("airflow.models.dagbag.DagBag.collect_dags")
-    @patch("airflow.models.serialized_dag.SerializedDagModel.write_dag")
-    @patch("airflow.models.dag.DAG.bulk_write_to_db")
-    def test_sync_to_db_is_retried(self, mock_bulk_write_to_db, mock_s10n_write_dag, mock_collect_dags):
```
Move to test_collection

ashb commented on tests/models/test_dagbag.py:
```diff
@@ -655,154 +607,6 @@ def f():
-            mock_sync_perm_for_dag.assert_called_once_with(dag, session=session)
-
-    @patch("airflow.www.security_appless.ApplessAirflowSecurityManager")
-    def test_sync_perm_for_dag(self, mock_security_manager):
```
Moved to test_collection

ashb commented on tests/models/test_dagbag.py:
```diff
@@ -655,154 +607,6 @@ def f():
-            )
-
-    @patch("airflow.www.security_appless.ApplessAirflowSecurityManager")
-    def test_sync_perm_for_dag_with_dict_access_control(self, mock_security_manager):
```
This was not moved --it was basically the same as test_sync_perm_for_dag but with a different format for the perms -- that is handled inside DAG though, so not relevant to test again here.

kaxil commented on airflow/dag_processing/collection.py:
```diff
@@ -163,6 +168,178 @@ def _update_dag_owner_links(dag_owner_links: dict[str, str], dm: DagModel, *, se
     )
 
 
+def _serialize_dag_capturing_errors(dag, session, processor_subdir):
```
Worth adding type hints

kaxil commented on airflow/dag_processing/collection.py:
```diff
@@ -163,6 +168,178 @@ def _update_dag_owner_links(dag_owner_links: dict[str, str], dm: DagModel, *, se
+    except OperationalError:
+        raise
+    except Exception:
+        log.exception("Failed to write serialized DAG: %s", dag.fileloc)
```
Do we want to log dag id, dag file path or both? the line says DAG but param is fileloc

kaxil commented on airflow/dag_processing/collection.py:
```diff
@@ -163,6 +168,178 @@ def _update_dag_owner_links(dag_owner_links: dict[str, str], dm: DagModel, *, se
+                # everything?
+            )
+        )
+    )
```
I don't think it is safe. The other type is `ASSET_CONFLICT` which is handled at a different place and should not go away with the file

kaxil replied:
https://github.com/apache/airflow/blob/224a4e3a2fbc585840aa54e5a5048ea285c4e179/airflow/jobs/scheduler_job_runner.py#L2246-L2270

ashb replied:
Hmmm yes. I wonder how to handle that, I don't really want this function to be type specific.... I guess it might have to take an optional list of DagWarningType, and if so only looks at those

kaxil commented on airflow/dag_processing/collection.py:
```diff
@@ -163,6 +168,178 @@ def _update_dag_owner_links(dag_owner_links: dict[str, str], dm: DagModel, *, se
+    from airflow.listeners.listener import get_listener_manager
+
+    # We can remove anything from files parsed in this batch that doesn't have an error. We need to remove old
+    # errors (i.e. form files that are removed) separately
```
```suggestion
    # errors (i.e. from files that are removed) separately
```

kaxil commented on tests/api_connexion/endpoints/test_task_endpoint.py:
```diff
@@ -86,7 +86,7 @@ def setup_dag(self, configured_app):
             mapped_dag.dag_id: mapped_dag,
             unscheduled_dag.dag_id: unscheduled_dag,
         }
-        DagBag._sync_to_db(dag_bag.dags)
```
:D

kaxil commented on tests/dag_processing/test_collection.py:
```diff
@@ -129,3 +154,235 @@ def test_add_asset_trigger_references(self, is_active, is_paused, expected_num_t
+
+        sync_perms_spy = spy_agency.spy_on(
+            airflow.dag_processing.collection._sync_dag_perms,
+            call_original=False,
```
TIL -- so this is like an actual mock

ashb committed:
Move ParseImportError formatting/depth tests to dagbag

This is a property of dagbag parsing, not the processor.

Also convert it to a parameterizes test too

ashb commented on tests/dag_processing/test_processor.py:
```diff
@@ -85,9 +78,7 @@ class TestDagFileProcessor:
         clear_db_runs()
-        clear_db_pools()
         clear_db_dags()
-        clear_db_import_errors()
```
These two aren't needed anymore as the tests that assert these don't live in here anymore - either in dagbag or processor tests

kaxil commented on tests/dag_processing/test_processor.py:
```diff
@@ -246,183 +243,6 @@ def test_process_file_should_failure_callback(self, monkeypatch, tmp_path, get_t
-
-            session.rollback()
-
-    def test_no_import_errors_with_parseable_dag_in_zip(self, tmp_path):
```
Worth keeping this -- for zipfile or do we check that somewhere else too?

ashb replied:
There's plenty of tests of zipfile parsing in dagbag, I'll add `        assert not dagbag.import_errors` to one of them

ashb committed:
fixup! Move ParseImportError formatting/depth tests to dagbag

ashb committed:
fixup! Move ParseImportError formatting/depth tests to dagbag
